src/distribution.cpp

```diff
@@ -0,0 +1,312 @@
+
+namespace openmc {
+
+//==============================================================================
+// Discrete implementation
+//==============================================================================
+
+Discrete::Discrete(pugi::xml_node node)
+{
+  auto params = get_node_array<double>(node, "parameters");
+
+  std::size_t n = params.size();
+  std::copy(params.begin(), params.begin() + n/2, std::back_inserter(x_));
+  std::copy(params.begin() + n/2, params.end(), std::back_inserter(p_));
+
+  normalize();
+}
+
+Discrete::Discrete(const double* x, const double* p, int n)
+  : x_{x, x+n}, p_{p, p+n}
+{
+  normalize();
+}
+
+double Discrete::sample() const
+{
+  int n = x_.size();
+  if (n > 1) {
+    double xi = prn();
+    double c = 0.0;
+    for (int i = 0; i < n; ++i) {
+      c += p_[i];
+      if (xi < c) return x_[i];
+    }
+    throw std::runtime_error{"Error when sampling probability mass function."};
+  } else {
+    return x_[0];
+  }
+}
+
+void Discrete::normalize()
+{
+  // Renormalize density function so that it sums to unity
+  double norm = std::accumulate(p_.begin(), p_.end(), 0.0);
+  for (auto& p_i : p_)
+    p_i /= norm;
+}
+
+//==============================================================================
+// Uniform implementation
+//==============================================================================
+
+Uniform::Uniform(pugi::xml_node node)
+{
+  auto params = get_node_array<double>(node, "parameters");
+  if (params.size() != 2)
+    openmc::fatal_error("Uniform distribution must have two "
+                        "parameters specified.");
+
+  a_ = params.at(0);
+  b_ = params.at(1);
+}
+
+double Uniform::sample() const
+{
+  return a_ + prn()*(b_ - a_);
+}
+
+//==============================================================================
+// Maxwell implementation
+//==============================================================================
+
+Maxwell::Maxwell(pugi::xml_node node)
+{
+  theta_ = std::stod(get_node_value(node, "parameters"));
+}
+
+double Maxwell::sample() const
+{
+  return maxwell_spectrum(theta_);
+}
+
+//==============================================================================
+// Watt implementation
+//==============================================================================
+
+Watt::Watt(pugi::xml_node node)
+{
+  auto params = get_node_array<double>(node, "parameters");
+  if (params.size() != 2)
+    openmc::fatal_error("Watt energy distribution must have two "
+                        "parameters specified.");
+
+  a_ = params.at(0);
+  b_ = params.at(1);
+}
+
+double Watt::sample() const
+{
+  return watt_spectrum(a_, b_);
+}
+
+//==============================================================================
+// Normal implementation
+//==============================================================================
+Normal::Normal(pugi::xml_node node) 
+{
+  auto params = get_node_array<double>(node,"parameters");
+  if (params.size() != 2)
```

Comment:
Nit: The established pattern in these files is to use { } even if only a single statement.